documents/read_acls.go

```diff
@@ -0,0 +1,336 @@
+package documents
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/centrifuge/centrifuge-protobufs/gen/go/coredocument"
+	"github.com/centrifuge/go-centrifuge/errors"
+	"github.com/centrifuge/go-centrifuge/identity"
+	"github.com/centrifuge/go-centrifuge/utils"
+	"github.com/centrifuge/precise-proofs/proofs/proto"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+)
+
+// initReadRules initiates the read rules for a given CoreDocumentModel.
+// Collaborators are given Read_Sign action.
+// if the rules are created already, this is a no-op.
+// if collaborators are empty, it is a no-op
+func (cd *CoreDocument) initReadRules(collaborators []identity.CentID) {
+	if len(cd.Document.Roles) > 0 && len(cd.Document.ReadRules) > 0 {
+		return
+	}
+
+	if len(collaborators) < 1 {
+		return
+	}
+
+	cd.addCollaboratorsToReadSignRules(collaborators)
+}
+
+// findRole calls OnRole for every role that matches the actions passed in
+func findRole(cd coredocumentpb.CoreDocument, onRole func(rridx, ridx int, role *coredocumentpb.Role) bool, actions ...coredocumentpb.Action) bool {
+	am := make(map[int32]struct{})
+	for _, a := range actions {
+		am[int32(a)] = struct{}{}
+	}
+
+	for i, rule := range cd.ReadRules {
+		if _, ok := am[int32(rule.Action)]; !ok {
+			continue
+		}
+
+		for j, rk := range rule.Roles {
+			role, err := getRole(rk, cd.Roles)
+			if err != nil {
+				// seems like roles and rules are not in sync
+				// skip to next one
+				continue
+			}
+
+			if onRole(i, j, role) {
+				return true
+			}
+
+		}
+	}
+
+	return false
+}
+
+// GetExternalCollaborators returns collaborators of a Document without the own centID.
+func (cd *CoreDocument) GetExternalCollaborators(self identity.CentID) ([][]byte, error) {
+	var cs [][]byte
+	for _, c := range cd.Document.Collaborators {
+		c := c
+		id, err := identity.ToCentID(c)
+		if err != nil {
+			return nil, errors.New("failed to convert to CentID: %v", err)
+		}
+		if !self.Equal(id) {
+			cs = append(cs, c)
+		}
+	}
+
+	return cs, nil
+}
+
+// NFTOwnerCanRead checks if the nft owner/account can read the Document
+func (cd *CoreDocument) NFTOwnerCanRead(tokenRegistry TokenRegistry, registry common.Address, tokenID []byte, account identity.CentID) error {
+	// check if the account can read the doc
+	if cd.AccountCanRead(account) {
+		return nil
+	}
+
+	// check if the nft is present in read rules
+	found := findRole(cd.Document, func(_, _ int, role *coredocumentpb.Role) bool {
+		_, found := isNFTInRole(role, registry, tokenID)
+		return found
+	}, coredocumentpb.Action_ACTION_READ)
```

Comment:
you can use the Action constant here (ACLRead)

Reply:
Lets be explict